tests/mod_learningmap_mapworker_test.php

```diff
@@ -0,0 +1,98 @@
+ * @group      mebis
+ * @covers     \mod_learningmap\mapworker
+ */
+class mod_learningmap_mapworker_test extends \advanced_testcase {
+
+    /**
+     * Prepare testing environment
+     */
+    public function setUp(): void {
+        global $DB;
+        $this->course = $this->getDataGenerator()->create_course(['enablecompletion' => 1]);
+        $this->learningmap = $this->getDataGenerator()->create_module('learningmap',
+            ['course' => $this->course, 'completion' => 2, 'completiontype' => 0]);
+
+        $this->activities = [];
+        for ($i = 0; $i < 9; $i++) {
+            $this->activities[] = $this->getDataGenerator()->create_module(
+                'page',
+                ['name' => 'A', 'content' => 'B', 'course' => $this->course, 'completion' => 2, 'completionview' => 1]
+            );
+            $this->learningmap->placestore = str_replace(99990 + $i, $this->activities[$i]->cmid, $this->learningmap->placestore);
+        }
+        $DB->set_field('learningmap', 'placestore', $this->learningmap->placestore, ['id' => $this->learningmap->id]);
+
+        $this->user1 = $this->getDataGenerator()->create_user(
+            [
+                'email' => 'user1@example.com',
+                'username' => 'user1'
+            ]
+        );
+
+        $this->modinfo = get_fast_modinfo($this->course, $this->user1->id);
+        $this->completion = new \completion_info($this->modinfo->get_course());
+        $this->cm = $this->modinfo->get_cm($this->learningmap->cmid);
+    }
+
+    /**
+     * Tests slicemode
+     *
+     * @return void
+     */
+    public function test_slicemode() : void {
+        $this->resetAfterTest();
+        $this->setAdminUser();
+        $this->setUser($this->user1);
+        $placestore = json_decode($this->learningmap->placestore, true);
+        $placestore['slicemode'] = true;
+        $mapworker = new mapworker($this->learningmap->intro, $placestore, $this->cm, false);
+        $mapworker->process_map_objects();
+        $expectedvalues = [
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 338 47 L 338 108 L 72 108 Z',
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 338 47 L 338 242 L 72 242 Z',
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 481 47 L 481 349 L 72 349 Z',
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 481 47 L 481 349 L 72 349 Z',
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 481 47 L 481 349 L 72 349 Z',
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 481 47 L 481 349 L 72 349 Z',
+            'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 72 47 L 649 47 L 649 349 L 72 349 Z',
+            null
+        ];
+        $overlay = $mapworker->get_attribute('learningmap-overlay', 'd');
+        $this->assertEquals($overlay, 'M 0 0 L 0 2111 L 800 2111 L 800 0 Z M 37 12 L 137 12 L 137 112 L 37 112 Z');
+
+        for ($i = 0; $i < 8; $i++) {
+            $acm = $this->modinfo->get_cm($this->activities[$i]->cmid);
```

Comment:
Can the name of `$acm` be improved? What does "acm" mean?

Reply:
abbreviation for activity course module, changed now